internal/controller/argocdcommitstatus_controller.go

```diff
@@ -0,0 +1,398 @@
+	repo         string
+	targetBranch string
+}
+
+// ArgoCDCommitStatusReconciler reconciles a ArgoCDCommitStatus object
+type ArgoCDCommitStatusReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	PathLookup utils.PathLookup
+}
+
+// +kubebuilder:rbac:groups=promoter.argoproj.io,resources=argocdcommitstatuses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=promoter.argoproj.io,resources=argocdcommitstatuses/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=promoter.argoproj.io,resources=argocdcommitstatuses/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the ArgoCDCommitStatus object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.1/pkg/reconcile
+func (r *ArgoCDCommitStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling ArgoCDCommitStatus")
+	var argoCDCommitStatus promoterv1alpha1.ArgoCDCommitStatus
+	err := r.Get(ctx, req.NamespacedName, &argoCDCommitStatus, &client.GetOptions{})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			logger.Info("ArgoCDCommitStatus not found")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "failed to get ArgoCDCommitStatus")
+		return ctrl.Result{}, fmt.Errorf("failed to get ArgoCDCommitStatus: %w", err)
+	}
+
+	aggregates := map[objKey][]*Aggregate{}
+
+	ls, err := metav1.LabelSelectorAsSelector(argoCDCommitStatus.Spec.ApplicationSelector)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to parse label selector: %w", err)
+	}
+	// TODO: we should setup a field index and only list apps related to the currently reconciled app
+	var ul unstructured.UnstructuredList
+	ul.SetGroupVersionKind(gvk)
+	err = r.Client.List(ctx, &ul, &client.ListOptions{
+		LabelSelector: ls,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list CommitStatus objects: %w", err)
+	}
+
+	gitAuthProvider, repositoryRef, err := r.getGitAuthProvider(ctx, argoCDCommitStatus)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get git auth provider: %w", err)
+	}
+
+	argoCDCommitStatus.Status.ApplicationsSelected = []promoterv1alpha1.ApplicationsSelected{}
+	for _, obj := range ul.Items {
+		var application argocd.ArgoCDApplication
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &application)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to cast unstructured object to typed object: %w", err)
+		}
+
+		if application.Spec.SourceHydrator == nil {
+			continue
+		}
```

Comment:
We should document that this currently only supports apps using the hydrator, even though you could theoretically use promoter without Argo CD's hydrator.